config/authsources.php

```diff
@@ -37,10 +37,9 @@
 $config = [];
 
 // Case for specifying no $SESSION IdP, select the first configured IdP as the default.
-$arr = array_reverse($saml2auth->metadataentities);
-$metadataentities = array_pop($arr);
-$idpentity = array_pop($metadataentities);
-$idp = md5($idpentity->entityid);
+$metadataentities = reset($saml2auth->metadataentities);
+$idpentity = reset($metadataentities);
+$idp = $idpentity->entityid;
```

Comment:
While this method works, I'm pretty sure this will fail in scenarios where there are two or more idp's configured and it will always choose the first if the session var is empty and it won't align and you'll get a redirect loop or some sort of failure. I'd like for that scenario to be tested and confirmed one way or another before this lands.

Also I'd like this logic refactored out because it is duplicated here which also needs to be updated:

https://github.com/catalyst/moodle-auth_saml2/blob/master/auth.php#L458-L465

It also seems duplicated by this logic here:

https://github.com/catalyst/moodle-auth_saml2/blob/master/auth.php#L488

It seems like an ideal place for it to go would be here:

https://github.com/catalyst/moodle-auth_saml2/blob/master/locallib.php#L452

Buuuutt ... all of this still feels wrong. If you have 2 idps, and the one you want is not the first, and the session has expired then it will never work. This stuff probably never needs to be in the session in the first place. If it was simply encoded into a query param in the saml returnto url then I think the whole problem is side stepped properly:

https://github.com/catalyst/moodle-auth_saml2/blob/master/auth.php#L507

```
$auth->requireAuth([
    'ReturnTo' => $ME + 'idp=entityid',

```
https://simplesamlphp.org/docs/stable/simplesamlphp-sp-api#section_3